kafka/src/main/java/org/togglz/kafka/KafkaStateRepository.java

```diff
@@ -0,0 +1,495 @@
+ * between them. The inbound topic should be configured to use the cleanup policy <i>compact</i>
+ * since, on the hand, it is used as persistent storage and, on the other, it has to be entirely
+ * consumed during the initialization of the {@link KafkaStateRepository}.
+ *
+ * @author Florian Stefan
+ */
+public class KafkaStateRepository implements AutoCloseable, StateRepository {
+
+  private static final Logger LOG = LoggerFactory.getLogger(KafkaStateRepository.class);
+
+  private final Map<String, FeatureStateStorageWrapper> featureStates;
+  private final FeatureStateConsumer featureStateConsumer;
+  private final FeatureStateProducer featureStateProducer;
+
+  private KafkaStateRepository(Builder builder) {
+    featureStates = new ConcurrentHashMap<>();
+    featureStateConsumer = new FeatureStateConsumer(builder, featureStates::put);
+    featureStateProducer = new FeatureStateProducer(builder);
+
+    featureStateConsumer.start();
+  }
+
+  /**
+   * Returns a new instance of a {@link KafkaStateRepository} builder.
+   *
+   * @return the new {@link KafkaStateRepository} builder
+   */
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  @Override
+  public void close() {
+    featureStateConsumer.close();
+    featureStateProducer.close();
+  }
+
+  @Override
+  public FeatureState getFeatureState(Feature feature) {
+    if (isRunning()) {
+      FeatureStateStorageWrapper storageWrapper = featureStates.get(feature.name());
+
+      if (storageWrapper != null) {
+        return featureStateForWrapper(feature, storageWrapper);
+      } else {
+        LOG.warn("Could not find featureState for given feature - fallback to default value.");
+        return null;
+      }
+    } else {
+      LOG.warn("FeatureStateConsumer not running anymore - fallback to default value.");
+      return null;
+    }
+  }
+
+  @Override
+  public void setFeatureState(FeatureState featureState) {
+    featureStateProducer.send(featureState.getFeature(), wrapperForFeatureState(featureState));
+  }
+
+  /**
+   * Returns {@code true} if the underlying Kafka consumer is running.
+   * If the Kafka consumer is not running, the {@link FeatureState} cache
+   * is not updated and {@link #getFeatureState(Feature)} might return stale
+   * data.
+   *
+   * @return {@code true} if the underlying Kafka consumer is running.
+   */
+  public boolean isRunning() {
+    return featureStateConsumer.isRunning();
+  }
+
+  /**
+   * Returns the current lag of the underlying Kafka consumer. If the lag is
+   * greater than zero, the {@link FeatureState} cache might contain stale data.
+   *
+   * @return the current consumer lag
+   */
+  public long consumerLag() {
+    return featureStateConsumer.consumerLag();
+  }
+
+  public static class Builder {
+
+    private String bootstrapServers;
+    private String inboundTopic;
+    private String outboundTopic;
+    private Duration pollingTimeout;
+    private Duration initializationTimeout;
```

Comment:
Would it make sense to define some useful defaults for `pollingTimeout` and `initializationTimeout`?

Reply:
Done.